internal/core/metadata/rest_deviceprofile.go

```diff
@@ -97,119 +101,33 @@ func restUpdateDeviceProfile(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	// Check if the Device Profile exists
-	var to models.DeviceProfile
-	// First try with ID
-	to, err := dbClient.GetDeviceProfileById(from.Id)
+	op := profile.NewUpdateDeviceProfileExecutor(dbClient, from)
+	dp, err := op.Execute()
 	if err != nil {
-		// Try with name
-		to, err = dbClient.GetDeviceProfileByName(from.Name)
-		if err != nil {
-			LoggingClient.Error(err.Error())
+		LoggingClient.Error(err.Error())
+		switch err.(type) {
+		case errors2.ErrDeviceProfileNotFound:
 			http.Error(w, err.Error(), http.StatusNotFound)
-			return
+		case *errors2.ErrDuplicateName:
+			http.Error(w, err.Error(), http.StatusConflict)
+		case errors2.ErrDeviceProfileInvalidState:
+			http.Error(w, err.Error(), http.StatusConflict)
+		default:
+			http.Error(w, err.Error(), http.StatusInternalServerError)
 		}
-	}
 
-	// Update the device profile fields based on the passed JSON
-	if err := updateDeviceProfileFields(from, &to, w); err != nil {
-		LoggingClient.Error(err.Error())
-		return
-	}
-	if err := dbClient.UpdateDeviceProfile(to); err != nil {
-		LoggingClient.Error(err.Error())
-		http.Error(w, err.Error(), http.StatusInternalServerError)
 		return
 	}
 
 	// Notify Associates
-	notifyProfileAssociates(to, http.MethodPut)
-
-	w.Header().Set("Content-Type", "application/json")
-	w.WriteHeader(http.StatusOK)
-	w.Write([]byte("true"))
-}
-
-// Update the fields of the device profile
-// to - the device profile that was already in Mongo (whose fields we're updating)
-// from - the device profile that was passed in with the request
-func updateDeviceProfileFields(from models.DeviceProfile, to *models.DeviceProfile, w http.ResponseWriter) error {
-	if from.Description != "" {
-		to.Description = from.Description
-	}
-	if from.Labels != nil {
-		to.Labels = from.Labels
-	}
-	if from.Manufacturer != "" {
-		to.Manufacturer = from.Manufacturer
-	}
-	if from.Model != "" {
-		to.Model = from.Model
-	}
-	if from.Origin != 0 {
-		to.Origin = from.Origin
-	}
-	if from.Name != "" {
-		to.Name = from.Name
-		// Names must be unique for each device profile
-		if err := checkDuplicateProfileNames(*to, w); err != nil {
-			return err
-		}
-	}
-	if from.DeviceResources != nil {
-		to.DeviceResources = from.DeviceResources
-	}
-	if from.DeviceCommands != nil {
-		to.DeviceCommands = from.DeviceCommands
-	}
-	if from.CoreCommands != nil {
-		// Check for duplicates by command name
-		if err := checkDuplicateCommands(from, w); err != nil {
-			return err
-		}
-		to.CoreCommands = from.CoreCommands
-	}
-
-	return nil
-}
-
-// Check for duplicate names in device profiles
-func checkDuplicateProfileNames(dp models.DeviceProfile, w http.ResponseWriter) error {
-	profiles, err := dbClient.GetAllDeviceProfiles()
+	err = notifyProfileAssociates(dp, dbClient, http.MethodPut)
```

Comment:
I did not include tests to cover the `notifyProfileAssociates` function. This function is tightly coupled to the `http` client. I did not want to complicate this PR anymore by refactoring further to decouple the HTTP client, so I left those tests out.

Reply:
Yep, that's fine. When I created the original test for notifications in the AddNewDevice path, I knew there would be other areas that would need a similar abstraction. Could be they'll all converge at some point.